### PR TITLE
Expirable key map refactor

### PR DESCRIPTION
**Pull request by mendes4212**
This PR aim to correct the behavior that create multiples queries plan with same script only changing expiring Data , and this was overloading the database, the new behavior correct creating a single query plan reusing a ExpirationDate.

### CODE REVIEW

File: src/Take.Elephant.Sql/DbConnectionDecorator.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Take.Elephant.Sql
+{
+    public sealed class DbConnectionDecorator : DbConnection
```
Comment (andreminelli) [inline src/Take.Elephant.Sql/DbConnectionDecorator.cs:10]: The class name needs a more specific name, like ExpirationDbConnection.

And it will be better putting this class inside ExpirableKeySqlMap, since it is supposed to be used only there, and only in conjunction with ExpirationDatabaseDriver - those 2 decorators are an internal design detail, not supposed to be publicly available. 

Another reason to put them 'close' is that both must use the same parameter name, and both should use a single constant value for this. 

In summary, I suggest:
-Rename this class to ExpirationDbConnection
-Move it inside ExpirableKeySqlMap
-Make this class and ExpirationDatabaseDriver `internal` (instead of public or private)
-Use attribute InternalsVisibleTo to still allow unit test project to access this class (to keep things simple)

Comment (mendes4212) [reply src/Take.Elephant.Sql/DbConnectionDecorator.cs]: Thanks for the sugestions , I implemented them all.

File: src/Take.Elephant.Tests/ExpirableKeyMapFacts.cs [hunk 2]
```diff
@@ -24,7 +24,7 @@ public virtual TimeSpan CreateTtl()
         {
             return TimeSpan.FromMilliseconds(500);
         }
-
+
```
Comment (gabrielgfaria) [inline src/Take.Elephant.Tests/ExpirableKeyMapFacts.cs:27]: Remove superfluous spacing

Comment (mendes4212) [reply src/Take.Elephant.Tests/ExpirableKeyMapFacts.cs]: I didn't see it, thanks, removed

**Review by gabrielgfaria — APPROVED** [review]

File: src/Take.Elephant.Sql/ExpirableKeySqlMap.cs [hunk 3]
```diff
@@ -157,4 +163,69 @@ private static string InjectSqlFilter(string sql, string filter)
...
+            return command;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+            _dbConnection.Dispose();
+        }
+
+        public override async ValueTask DisposeAsync()
```
Comment (andreminelli) [inline src/Take.Elephant.Sql/ExpirableKeySqlMap.cs:220]: public method should come first.

Comment (mendes4212) [reply src/Take.Elephant.Sql/ExpirableKeySqlMap.cs]: Done

File: src/Take.Elephant.Sql/ExpirableKeySqlMap.cs [hunk 4]
```diff
@@ -157,4 +163,69 @@ private static string InjectSqlFilter(string sql, string filter)
...
+        public override void Open()
+        {
+            _dbConnection.Open();
+        }
+
+        protected override DbCommand CreateDbCommand()
+        {
+            var command = _dbConnection.CreateCommand();
+            var parameter = command.CreateParameter();
+            parameter.ParameterName = "@ExpirableKeySqlMap_ExpirationDate";
```
Comment (andreminelli) [inline src/Take.Elephant.Sql/ExpirableKeySqlMap.cs:213]: Extract this name to a constant, that ExpirationDatabaseDriver can use when injecting its filter.

Comment (mendes4212) [reply src/Take.Elephant.Sql/ExpirableKeySqlMap.cs]: Done

File: src/Take.Elephant.Tests/Sql/DbConnectionDecoratorFacts.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+﻿using Take.Elephant.Sql;
+using Xunit;
+using NSubstitute;
+using System.Data.Common;
+using System.Threading.Tasks;
+
+namespace Take.Elephant.Tests.Sql
+{
+    public class DbConnectionDecoratorFacts : FactsBase
```
Comment (andreminelli) [inline src/Take.Elephant.Tests/Sql/DbConnectionDecoratorFacts.cs:9]: This class name should be change since the related class has changed.

File: src/Take.Elephant.Sql/ExpirableKeySqlMap.cs [hunk 6]
```diff
@@ -157,4 +161,69 @@ private static string InjectSqlFilter(string sql, string filter)
...
+        {
+            await base.DisposeAsync();
+            await _dbConnection.DisposeAsync();
+        }
+
+        protected override DbCommand CreateDbCommand()
+        {
+            var command = _dbConnection.CreateCommand();
+            var parameter = command.CreateParameter();
+            parameter.ParameterName = ExpirableKeySqlMap<int,int>.ExpirationDatabaseDriverDecorator.EXPIRATION_DATE_PARAMETER_NAME;
```
Comment (andreminelli) [inline src/Take.Elephant.Sql/ExpirableKeySqlMap.cs:211]: Putting this class inside ExpirabkeKeySqlMap - as ExpirationDatabaseDriverDecorator is - will make more clear that there are all related, and allow to remove the `ExpirableKeySqlMap<int,int>` here.

File: src/Take.Elephant.Tests/Sql/DbConnectionDecoratorFacts.cs (hunk 5 above)
Comment (mendes4212) [reply src/Take.Elephant.Tests/Sql/DbConnectionDecoratorFacts.cs]: I really forgot, done

File: src/Take.Elephant.Sql/ExpirableKeySqlMap.cs (hunk 6 above)
Comment (mendes4212) [reply src/Take.Elephant.Sql/ExpirableKeySqlMap.cs:211]: But if I do it , I'll have to use `ExpirableKeySqlMap<int,int>` on test class.

Comment (andreminelli) [reply src/Take.Elephant.Sql/ExpirableKeySqlMap.cs:211]: Yes, but I think it's better there than here...

Comment (mendes4212) [reply src/Take.Elephant.Sql/ExpirableKeySqlMap.cs]: Ok, done
